swarm/src/handler.rs

```diff
@@ -338,94 +335,110 @@ impl<'a> ProtocolsChange<'a> {
     ///
     /// Returns `None` if the change is a no-op, i.e. `to_add` is a subset of `existing_protocols`.
     pub(crate) fn add(
-        existing_protocols: &'a HashSet<StreamProtocol>,
-        to_add: &'a HashSet<StreamProtocol>,
+        to_add: HashSet<StreamProtocol>,
+        existing_protocols: &HashSet<StreamProtocol>,
+        temp_protocols: &'a mut Vec<StreamProtocol>,
     ) -> Option<Self> {
-        let mut actually_added_protocols = to_add.difference(existing_protocols).peekable();
-
-        actually_added_protocols.peek()?;
+        temp_protocols.clear();
+        temp_protocols.extend(
+            to_add
+                .into_iter()
+                .filter(|i| !existing_protocols.contains(i)),
+        );
+
+        if temp_protocols.is_empty() {
+            return None;
+        }
 
-        Some(ProtocolsChange::Added(ProtocolsAdded {
-            protocols: actually_added_protocols,
+        Some(Self::Added(ProtocolsAdded {
+            protocols: temp_protocols.iter(),
         }))
     }
 
-    /// Compute the [`ProtocolsChange`] that results from removing `to_remove` from `existing_protocols`.
+    /// Compute the [`ProtocolsChange`] that results from removing `to_remove` from `existing_protocols`. Removes the protocols from `existing_protocols`.
     ///
     /// Returns `None` if the change is a no-op, i.e. none of the protocols in `to_remove` are in `existing_protocols`.
     pub(crate) fn remove(
-        existing_protocols: &'a HashSet<StreamProtocol>,
-        to_remove: &'a HashSet<StreamProtocol>,
+        to_remove: HashSet<StreamProtocol>,
+        existing_protocols: &mut HashSet<StreamProtocol>,
```

Comment:
If we are still accepting two `HashSet`s, why can we not continue to use the `HashSet` combinators? How is the new implementation more performant than the old one?

Reply:
Same question applies to `add`.

Reply:
Because of the change on `ProtocolChange` I cannot use `difference` without making the type more complex, I also thought this would be more consistent.